graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,372 @@
+ */
+public class CacheControlInstrumentation extends SimpleInstrumentation {
+  private final int defaultMaxAge;
+  private final HintCache hintCache = new HintCache();
+
+  private static final String DIRECTIVE_NAME = "cacheControl";
+  private static final String EXTENSION_KEY = "cache-control";
+
+  public CacheControlInstrumentation() {
+    this(0);
+  }
+
+  public CacheControlInstrumentation(int defaultMaxAge) {
+    this.defaultMaxAge = defaultMaxAge;
+  }
+
+  @Override
+  public InstrumentationState createState() {
+    return new CacheControlState();
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginField(
+      InstrumentationFieldParameters parameters) {
+    CacheControlState state = parameters.getInstrumentationState();
+    CacheControlPolicy fieldPolicy = new CacheControlPolicy();
+    boolean inheritMaxAge = false;
+
+    GraphQLUnmodifiedType unwrappedReturnType =
+        GraphQLTypeUtil.unwrapAll(parameters.getExecutionStepInfo().getType());
+
+    // There's no way to set a cacheControl directive on the _entities field or the _Entity union in
+    // SDL.
+    // Instead, we can determine the possible concrete types from the representations arguments and
+    // select the
+    // most restrictive cache policy from those types.
+
+    if (unwrappedReturnType.getName().equals(_Entity.typeName)) {
+      Object representations = parameters.getExecutionStepInfo().getArgument(_Entity.argumentName);
+
+      if (representations instanceof List) {
+        typesFromEntitiesArgument(
+                representations, parameters.getExecutionContext().getGraphQLSchema())
+            .stream()
+            .map(type -> hintCache.getHint((GraphQLCompositeType) type))
+            .filter(Optional::isPresent)
+            .forEach(hint -> fieldPolicy.restrict(hint.get()));
+      }
+    }
+
+    // Cache hint on the return type of this field if it's a composite type
+
+    if (unwrappedReturnType instanceof GraphQLCompositeType) {
+      Optional<Hint> hint = hintCache.getHint((GraphQLCompositeType) unwrappedReturnType);
+
+      if (hint.isPresent()) {
+        fieldPolicy.replace(hint.get());
+        inheritMaxAge = hint.get().getInheritMaxAge().orElse(false);
+      }
+    }
+
+    // Cache hint on the field itself
+
+    Optional<Hint> fieldHint = hintCache.getHint(parameters.getField());
+    if (fieldHint.isPresent()) {
+      Hint hint = fieldHint.get();
+
+      // If inheritMaxAge is true, take note of that to avoid setting the default max age in the
+      // next step.
+      // This does allow setting the cache scope though.
+      //
+      // Note: setting both maxAge and inheritMaxAge doesn't make sense.
+      if (hint.getInheritMaxAge().orElse(false) && !fieldPolicy.hasMaxAge()) {
+        inheritMaxAge = true;
+        fieldPolicy.replace(hint.getScope());
+      } else {
+        fieldPolicy.replace(hint);
```

Comment:
No, you don't want to reset here. For example, if the return type was annotated with `@cacheControl(inheritMaxAge: true)` and the field was annotated with `@cacheControl(scope: PRIVATE)` then you don't want to ignore the inheritMaxAge.

(And if the field was annotated with `@cacheControl(maxAge: 10)` then it doesn't matter what you leave `inheritMaxAge` as because we only care about the value of the local variable `inheritMaxAge` if max age isn't otherwise set.)

BTW the code this is porting is https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-core/src/plugin/cacheControl/index.ts

Reply:
what if you have this?

```graphql
type Cat @cacheControl(maxAge: 60) {
  meowww: String
}

type Query {
  cat: Cat @cacheControl(inheritMaxAge: true)
}
```

who wins?

Reply:
There is in fact a comment about this :)

```
              // Note that specifying `@cacheControl(inheritMaxAge: true)` on a
              // field whose return type defines a `maxAge` gives precedence to
              // the type's `maxAge`. (Perhaps this should be some sort of
              // error.)
```

(looks like you found it though)